src/index.js

```diff
@@ -0,0 +1,849 @@
+'use strict'
+
+const Pubsub = require('libp2p-pubsub')
+const FloodSub = require('libp2p-floodsub')
+
+const pull = require('pull-stream')
+const lp = require('pull-length-prefixed')
+const asyncEach = require('async/each')
+const nextTick = require('async/nextTick')
+
+const MessageCache = require('./messageCache').MessageCache
+const CacheEntry = require('./messageCache').CacheEntry
+const utils = require('./utils')
+const assert = require('assert')
+
+const RPC = require('./message').rpc.RPC
+const constants = require('./constants')
+
+
+class GossipSub extends Pubsub {
+
+    /**
+     * @param {Object} libp2p
+     * @constructor
+     */
+    constructor (libp2p) {
+        super('libp2p:gossipsub', constants.GossipSubID, libp2p)
+
+	/**
+	 * Map of topic meshes
+	 *
+	 * @type {Map<string, Set<Peer>>}
+	 */
+	this.mesh = new Map()
+
+	/**
+	 * Map of topics to lists of peers. These mesh peers are the ones to which we are publishing without a topic membership
+	 *
+	 *@type {Map<string, Set<Peer>>}
+	 */
+	this.fanout = new Map()
+
+	/**
+	 * Map of last publish time for fanout topics
+	 *
+	 *@type {Map<string, Number>}
+	 */
+	this.lastpub = new Map()
+	
+	/**
+	 * Map of pending messages to gossip
+	 *
+	 * @type {Map<Peer, Array<RPC.ControlIHave object>> }
+	 */
+	this.gossip = new Map()
+	
+	/**
+	 * Map of control messages
+	 *
+	 * @type {Map<Peer, RPC.ControlMessage object>}
+	 */
+	this.control = new Map()
+
+	/**
+	 * A message cache that contains the messages for last few hearbeat ticks
+	 *
+	 */
+	this.messageCache = new MessageCache(constants.GossipSubHistoryGossip, constants.GossipSubHistoryLength)
+	
+    }
+
+    /**
+     * Removes a peer from the router
+     * 
+     * @override
+     * @param {Peer} peer
+     * @returns {undefined}
+     */
+    _removePeer (peer) {
+        const id = peer.info.id.toB58String()
+	
+	this.log('remove', id, peer._references)
+	// Only delete when no one else if referencing this peer.
+	if (--peer._references === 0){
+	    this.log('delete peer', id)
+            this.peers.delete(id)
+	
+            // Remove this peer from the mesh
+            for(let [topic, peers] of this.mesh.entries()){
+	        peers.delete(peer)
+	    }
+	    // Remove this peer from the fanout
+            for (let [topic, peers] of this.fanout.entries()){
```

Comment:
since you're not using `topic` both here and on `line 89` I would change it to and underscore to identify you're not using that. `topic` -> `_`